### PR TITLE
Route 404 errors to the 404 handler

### PR DESCRIPTION
**Pull request by mcollina**
#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message and code follows [Code of conduct](https://github.com/fastify/fastify/blob/master/CODE_OF_CONDUCT.md)

As titled. Does this needs a doc change? I think this fixes a bug in our current error handling logic.

(I need this for #724 to fully work)

### CODE REVIEW

File: lib/reply.js [hunk 1]
```diff
@@ -193,11 +193,24 @@ function handleError (reply, error, cb) {
     statusCode = statusCode || 500
   } else if (error.status >= 400) {
     statusCode = error.status
+    if (error.status === 404) {
```
Comment (delvedor) [inline lib/reply.js]: Why duplicate this logic?
You can just check `statusCode` few lines below (before `if (statusCode < 400)`)

Comment (mcollina) [reply lib/reply.js]: It can't be done later, because https://github.com/fastify/fastify/blob/938f8a551b03735e5bafa81d4c6927d2f9615ed1/lib/reply.js#L191 set the `statusCode`  variable to 404 at the beginning, breaking a good chunk of our tests.

File: test/404s.test.js [hunk 2]
```diff
@@ -817,7 +834,7 @@ test('recognizes errors from the http-errors module', t => {
   const fastify = Fastify()
 
   fastify.get('/', function (req, reply) {
-    reply.send(httpErrors.NotFound())
+    reply.send(httpErrors.Forbidden())
```
Comment (allevo) [inline test/404s.test.js]: Why this change?

Comment (mcollina) [reply test/404s.test.js]: Now that I've read it better, I misunderstood the test.

Comment (mcollina) [reply test/404s.test.js]: updated, PTAL

File: test/404s.test.js [hunk 3]
```diff
@@ -94,6 +99,18 @@ test('customized 404', t => {
...
+
+    test('with error object', t => {
+      t.plan(3)
+      sget({
+        method: 'GET',
+        url: 'http://localhost:' + fastify.server.address().port + '/with-error'
+      }, (err, response, body) => {
+        t.error(err)
+        t.strictEqual(response.statusCode, 404)
+        t.strictEqual(body.toString(), 'this was not found')
```
Comment (allevo) [inline test/404s.test.js]: Why this request invokes the `setNotFoundHandler` ?

Comment (mcollina) [reply test/404s.test.js]: because the error has a `statusCode` 404, this is _exactly_ the change this PR introduces.

Comment (allevo) [reply test/404s.test.js]: > (I need this for #724 to fully work)

Why?

If developer needs that the output is handled by the 404 route he should call `reply.notFound()`.

Comment (mcollina) [reply test/404s.test.js]: https://github.com/fastify/fastify-static/blob/master/index.js#L34-L48 completely jumps over `reply.send()` and it avoids calling all of our hook system, just to call `reply.notFound()`. I think that's not acceptable, so here are #724 and #727.

Comment (allevo) [reply test/404s.test.js]: This PR can introduce some side effect.

Anyway I'm not against this PR

**Comment by nwoltman** [comment]
@mcollina What are your thoughts on removing `reply.notFound` after this PR and #724 are merged? That API was only created to solve the problem of calling the 404 handler in `fastify-static` and this PR fixes that problem. After this is merged there will be 2 ways of calling the 404 handler: 1) `reply.notFound` and 2) by sending a 404 error. There doesn't need to be 2 ways of doing the exact same thing.

**Comment by mcollina** [comment]
@nwoltman I think we might even remove it, should I?
(it's one less API, and this is more elegant)

**Comment by mcollina** [comment]
I've just removed `reply.notFound()`, check it out.

File: docs/Reply.md [hunk 4]
```diff
@@ -180,6 +163,19 @@ fastify.get('/', function (request, reply) {
 
 If you want to completely customize the error response, checkout [`setErrorHandler`](https://github.com/fastify/fastify/blob/error-docs/docs/Server-Methods.md#seterrorhandler) API.
 
+Errors with a `status` or `statusCode` property equal to `404` will be routed to the not found handler, checkout
+the [`server.setNotFoundHandler`](https://github.com/fastify/fastify/blob/error-docs/docs/Server-Methods.md#setnotfoundhandler) API to know how to modify that:
```
Comment (jsumners) [inline docs/Reply.md]: "handler. See the [`server.setNotFoundHandler`] API to learn more about handling such cases:"

**Comment by allevo** [comment]
> > NB: what happen if
> > ```js
> > fastify.setNotFoundHandler(function (req, reply) {
> >       reply.send(new errors.NotFound())
> > })
> > ```
> > Maybe there's a loop.

> @allevo In that case, the 404 handler would not be called again and the response would be treated like a normal error.

@mcollina Can you add a test for this case?

**Comment by mcollina** [comment]
The test is already in there.

**Comment by nwoltman** [comment]
Well the code currently sends a basic response if a 404 error is sent inside a not-found handler (instead of treating it like a normal error), but that's fine. User's shouldn't be doing that anyway.

**Comment by allevo** [comment]
> Well the code currently sends a basic response if a 404 error is sent inside a not-found handler (instead of treating it like a normal error), but that's fine. User's shouldn't be doing that anyway.

There's no test for that. @mcollina Could you add it?

```js
fastify.get('/', function (request, reply) {
  reply.code(404).send({hello: 'this should be ignored'})
})
```

**Comment by jsumners** [comment]
I think this PR is good to go once the documentation is tweaked as I suggested.

**Comment by mcollina** [comment]
@allevo 

```js
fastify.get('/', function (request, reply) {
  reply.code(404).send({hello: 'this is always sent'})
})
```

This will keep working as it does right now. What will be different is:

```js
fastify.get('/', function (request, reply) {
  var err = new Error('something is missing')
  err.statusCode = 404
  reply.send(err)
})
```

Which will go to the not found handler instead of the error handler.

**Comment by mcollina** [comment]
PTAL

**Comment by mcollina** [comment]
`onRequest` will skip this logic, as we do not have our Request and Reply objects yet.

I've added a test for `'onSend'`.

**Comment by nwoltman** [comment]
> `onRequest` will skip this logic, as we do not have our Request and Reply objects yet.

If an `onRequest` hook calls `next(httpErrors.NotFound())`, the not-found handler will be invoked because the Request and Reply objects get created to send the error:

https://github.com/fastify/fastify/blob/be55c07fd6d091353c0b7ddaaedccd66e2948e9e/fastify.js#L293-L301
